OpenRA.Mods.RA/Traits/Attack/AttackLeap.cs

```diff
@@ -9,49 +9,47 @@
  */
 #endregion
 
+using System;
 using System.Collections.Generic;
-using System.Linq;
+using OpenRA.Activities;
 using OpenRA.Mods.Common.Traits;
 using OpenRA.Mods.RA.Activities;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.RA.Traits
 {
-	[Desc("Dogs use this attack model.")]
-	class AttackLeapInfo : AttackFrontalInfo
+	[Desc("Move onto the target then execute the attack.")]
+	public class AttackLeapInfo : AttackFrontalInfo, Requires<MobileInfo>
 	{
 		[Desc("Leap speed (in units/tick).")]
 		public readonly WDist Speed = new WDist(426);
 
-		public readonly WAngle Angle = WAngle.FromDegrees(20);
-
 		public override object Create(ActorInitializer init) { return new AttackLeap(init.Self, this); }
 	}
 
-	class AttackLeap : AttackFrontal
+	public class AttackLeap : AttackFrontal
 	{
 		readonly AttackLeapInfo info;
+		readonly Mobile mobile;
 
 		public AttackLeap(Actor self, AttackLeapInfo info)
 			: base(self, info)
 		{
 			this.info = info;
+			mobile = self.Trait<Mobile>();
 		}
 
 		public override void DoAttack(Actor self, Target target, IEnumerable<Armament> armaments = null)
 		{
-			if (target.Type != TargetType.Actor || !CanAttack(self, target))
-				return;
-
-			var a = ChooseArmamentsForTarget(target, true).FirstOrDefault();
-			if (a == null)
-				return;
-
-			if (!target.IsInRange(self.CenterPosition, a.MaxRange()))
+			if (IsAttacking || !mobile.CanMoveFreelyInto(self.World.Map.CellContaining(target.CenterPosition)))
```

Comment:
This should be moving to the target infantry's sub-cell rather than the center of the cell.